src/plugins/AvatarPEPPlugin.ts

```diff
@@ -0,0 +1,245 @@
+import { AbstractPlugin, IMetaData } from '../plugin/AbstractPlugin';
+import PluginAPI from '../plugin/PluginAPI';
+import { IContact } from '../Contact.interface';
+import Avatar from '../Avatar';
+import JID from '../JID';
+import Translation from '@util/Translation';
+import Hash from '@util/Hash';
+import { IAvatar } from '@src/Avatar.interface';
+import AvatarUI from '../ui/AvatarSet';
+import FileHelper from '@util/FileHelper';
+
+const MIN_VERSION = '4.0.0';
+const MAX_VERSION = '99.0.0';
+
+const UANS_BASE = 'urn:xmpp:avatar';
+const UANS_METADATA = +UANS_BASE + ':metadata';
+const UANS_NOTIFY = UANS_METADATA + '+notify';
+const UANS_DATA = UANS_BASE + ':data';
+
+export default class AvatarPEPPlugin extends AbstractPlugin {
+   public static getId(): string {
+      return 'pep-avatars';
+   }
+
+   public static getName(): string {
+      return 'PEP-based Avatars';
+   }
+
+   public static getMetaData(): IMetaData {
+      return {
+         description: Translation.t('setting-pep-avatar-enable'),
+         xeps: [
+            {
+               id: 'XEP-0084',
+               name: 'User Avatar',
+               version: '1.1.4',
+            },
+         ],
+      };
+   }
+
+   constructor(pluginAPI: PluginAPI) {
+      super(MIN_VERSION, MAX_VERSION, pluginAPI);
+
+      let connection = pluginAPI.getConnection();
+
+      connection.registerHandler(this.onMessageAvatarUpdate, 'http://jabber.org/protocol/pubsub#event', 'message');
+
+      pluginAPI.addAvatarProcessor(this.avatarProcessor, 49);
+      pluginAPI.addPublishAvatarProcessor(this.publishAvatarProcessor, 49);
+      pluginAPI.addFeature(UANS_NOTIFY);
+   }
+
+   public getStorage() {
+      return this.pluginAPI.getStorage();
+   }
+
+   private publishAvatarProcessor = (avatar: IAvatar | null): Promise<[IAvatar]> => {
+      let connection = this.pluginAPI.getConnection();
+
+      if (!avatar || avatar.getData() === undefined) {
+         let item = $build('metadata', { xmlns: UANS_METADATA });
+
+         return connection
+            .getPEPService()
+            .publish(UANS_METADATA, item.tree(), UANS_METADATA)
+            .then(function (result) {
+               if ($(result).attr('type') === 'result') {
+                  return [undefined];
+               } else {
+                  return [avatar];
```

Comment:
Since there should be no avatar at this point, the if doesn't have any impact. Publish also resolves the promise only if the request was successful, therefore the type check is useless.

Reply:
removed check